cross_negative_script/cross_negative_script.gs

```diff
@@ -0,0 +1,216 @@
+/*
+    Name:           WeDiscover - Cross Negative Script
+    Description:    This script allows you to add cross negative keywords to campaigns or ad groups
+                    with the option to use the original match type or convert all to exact match.
+    License:        https://github.com/we-discover/public/blob/master/LICENSE
+    Version:        1.0.0
+    Released:       2024-10-16
+    Author:         Nathan Ifill (@nathanifill)
+    Contact:        scripts@we-discover.com
+*/
+
+/************************************************* SETTINGS **************************************************************/
+
+// Options to choose between campaign-level and ad group-level negative keywords
+const ADD_CAMPAIGN_LEVEL_NEGATIVES = true;
+const ADD_AD_GROUP_LEVEL_NEGATIVES = true;
+const MAX_NEGATIVE_KEYWORDS = 5000; // This determines the maximum number of negative keywords to add to a campaign or ad group
+
+// Option to choose negative keyword match type
+const USE_ORIGINAL_MATCH_TYPE = false; // Set to false to use negative exact match for all or true to keep original match type
+
+// Option to filter which campaigns to include or exclude by name. If you leave this blank, the script will work across all
+// campaigns and/or ad groups. Leaving both of these blank is not recommended as (depending on the size of your account)
+// the script may not complete before Google's 30-minute execution limit.
+const INCLUDE_CAMPAIGN_NAMES = ['ZA > Search']; // Example: ['IncludePart1', 'IncludePart2']
+const EXCLUDE_CAMPAIGN_NAMES = []; // Example: ['ExcludePart1', 'ExcludePart2']
+
+// Option to enable or disable logging for faster execution
+const ENABLE_LOGGING = false; // Set to true to log what changes the script is making or set to false to speed up the script
+
+/*************************************************************************************************************************/
+
+function main() {
+  log("Starting process...");
+  log(" ");
+  log("Add campaign-level negatives? " + ADD_CAMPAIGN_LEVEL_NEGATIVES);
+  log("Add ad group-level negatives? " + ADD_AD_GROUP_LEVEL_NEGATIVES);
+  log("Use original match type for negatives? " + USE_ORIGINAL_MATCH_TYPE);
+  log("");
+  if (INCLUDE_CAMPAIGN_NAMES.length > 0) {
+    log("Campaign name must contain: " + INCLUDE_CAMPAIGN_NAMES);
+  }
+  if (EXCLUDE_CAMPAIGN_NAMES.length > 0) {
+    log("Campaign name must not contain: " + EXCLUDE_CAMPAIGN_NAMES);
+  }
+
+  if (ADD_CAMPAIGN_LEVEL_NEGATIVES) {
```

Comment:
💭 If I was being picky I'd argue lines 47 through 129 have quite a lot of repetitive code that could be more generalised.

Not a required change by any means, but I would think it makes more sense to iterate through campaigns once, while you're in each campaign, check if you should process then add campaign negatives. If you're also processing ad group negatives, do so then. 

Alongside the reduction in code, it'll save double loading of the campaigns and checking their eligibility twice which  I'm sure will tangibly speed the process up -- which sounds like a core concern.

Reply:
I would think it's possible -- i.e. iterate through campaigns and then if you're changing them, do it. Then go to their ad groups, if you're changing it, do it. 

In any case, not a required change, just something to be aware of if timings are a concern.